[BEAM-4553] Implement graphite sink for MetricsPusher and refactor MetricsHttpSink test

echauchot opened the pull request:
This PR adds the Graphite sink to metrics pusher feature and also refactors MetricsHttpSink test.
I put changes in coherent separate isolable commits. They must not be squashed because they deal with different jira tickets.
R: @aromanenko-dev 
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

echauchot commented on runners/extensions-java/metrics/src/main/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSink.java:
```diff
@@ -0,0 +1,305 @@
+public class MetricsGraphiteSink implements MetricsSink {
+  private static final Charset UTF_8 = Charset.forName("UTF-8");
+  private static final Pattern WHITESPACE = Pattern.compile("[\\s]+");
+  private static final String SPACE_REPLACEMENT = "_";
```
I was just to give it a name, not for deduplication

aromanenko-dev replied:
If it's used only once then we could you use just character, but ok with a constant

echauchot commented on runners/extensions-java/metrics/src/main/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSink.java:
```diff
@@ -0,0 +1,305 @@
+    Iterable<MetricResult<DistributionResult>> distributions =
+        metricQueryResults.getDistributions();
+
+    for (MetricResult<Long> counter : counters) {
```
I thought about it but the difference between the loops are the generic type, so with type erasure, I would need to pass a class to the common method and use a switch class in the code. That would be even less maintainable

aromanenko-dev replied:
Yes, agree. In the same time, the code to append message is very similar, so you could extract it into such method:
```
  private void appendPayload(StringBuilder messagePayload, String message1, String message2) {
    try {
      messagePayload.append(message1);
    } catch (UnsupportedOperationException e) {
      if (!e.getMessage().contains("committed metrics")) {
        throw e;
      }
    }
    messagePayload.append(message2);
  }
```

echauchot replied:
+1

echauchot replied:
I put this try catch in the parent class of the message metric classes (cf comment bellow).

echauchot commented on runners/extensions-java/metrics/src/main/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSink.java:
```diff
@@ -0,0 +1,305 @@
+      long metricTimestamp, MetricResult<Long> counter, boolean committedValue) {
+    String metricMessage;
+    if (committedValue) {
+      metricMessage =
```
In the first version, it was done using StringBuilder but it was less readable with all the .append(string) that is why I replaced it with string format.

echauchot replied:
Same here it has all been refactored with the class hierarchy

echauchot commented on runners/extensions-java/metrics/src/test/java/org/apache/beam/runners/extensions/metrics/CustomMetricQueryResults.java:
```diff
@@ -0,0 +1,140 @@
+                  "This runner does not currently support committed"
+                      + " metrics results. Please use 'attempted' instead.");
+            }
+            return 10L;
```
yes just test value

echauchot commented on runners/extensions-java/metrics/src/test/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSinkTest.java:
```diff
@@ -0,0 +1,108 @@
+    pipelineOptions.setMetricsGraphiteHost("127.0.0.1");
+    MetricsGraphiteSink metricsGraphiteSink = new MetricsGraphiteSink(pipelineOptions);
+    metricsGraphiteSink.writeMetrics(metricQueryResults);
+    Thread.sleep(2000L);
```
Yes because, when we write messages to the socket, then on a different thread `NetworkMockServer` reads the socket and adds messages to an arrayList so that they could be read in the assert. On heavy load jenkins server, the test might fail because the `NetworkMockServer` does not have enough time to add the messages to the arraylist

aromanenko-dev replied:
Well, potentially it's still can be a cause of flaky tests - we never know how much time you need to wait. Could you avoid using two threads and socket in this case? Or use approach based on `CountDownLatch`?

echauchot replied:
+1

echauchot commented on runners/extensions-java/metrics/src/main/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSink.java:
```diff
@@ -0,0 +1,305 @@
+    return WHITESPACE.matcher(metricName).replaceAll(SPACE_REPLACEMENT);
+  }
+
+  private enum CommittedOrAttemped {
```
no, only two type of metrics

aromanenko-dev replied:
I was asking because these status names are incorporated into enum name, so it will be difficult to add another one if it's needed. I suggest to rename it with more generic name, like `MetricStatus` or similar one.

echauchot replied:
Take a look at this class: `AttemptedAndCommitted` in the SDK. I did this out of consistency, commited and attempted are the 2 types defined at beam model level.

aromanenko-dev replied:
Ok, let's keep it so (though, it' not a good practice in my opinion).

echauchot commented on runners/extensions-java/metrics/src/main/java/org/apache/beam/runners/extensions/metrics/MetricsGraphiteSink.java:
```diff
@@ -0,0 +1,305 @@
+              Locale.US,
+              "%s %s %s\n",
+              createNormalizedMetricName(
+                  distribution, "distribution", "min", CommittedOrAttemped.COMMITTED),
```
what do you mean by builder ? you mean string builder or refactoring the code to implement a builder pattern like` message.withType`, `message.withCommitted` etc...

aromanenko-dev replied:
The latter one - refactor to builder pattern.

echauchot replied:
I did the refactoring based on message objects. It is fine, it is readable and with a lot less code duplication in the String.format .

echauchot commented:
@aromanenko-dev thanks for the review ! I answered all your comments PTAL.

aromanenko-dev commented:
@echauchot I added my comments.

echauchot commented:
@aromanenko-dev Thanks for your suggestions, I have just finished a big refactoring to avoid code duplication. It is based now on a Metric Message class hierarchy. It addresses all your comments. I also replaced the Thread.sleep in all the metrics sinks tests (including already existing http metrics sink test) with a CountDownLatch to avoid flakiness of the tests in case of load. PTAL

fmendezh commented:
https://github.com/apache/beam/pull/6569/commits/0abb8bfeb6ea85b260226d2da1e7a869e698e074 addd Graphite specific methods, shouldn't that be on a new PipelineOptionsGraphiteMetrics class?
